rust/connlib/clients/shared/src/eventloop.rs

```diff
@@ -0,0 +1,467 @@
+                    res: ReplyMessages::SignedLogUrl(url),
+                    ..
+                }) => {
+                    let Some(path) = self.tunnel.callbacks().roll_log_file() else {
+                        continue;
+                    };
+
+                    tokio::spawn(async move {
+                        if let Err(e) = upload(path.clone(), url).await {
+                            tracing::warn!(
+                            "Failed to upload log file at path {path_display}: {e}. Not retrying.",
+                            path_display = path.display()
+                        );
+                        }
+                    });
+                }
+                Poll::Ready(phoenix_channel::Event::HeartbeatSent) => {}
+                Poll::Ready(phoenix_channel::Event::InboundReq { req_id, req }) => {}
+                Poll::Ready(phoenix_channel::Event::JoinedRoom { topic }) => {}
+                Poll::Ready(phoenix_channel::Event::Disconnect(reason)) => {
+                    return Poll::Ready(Err(anyhow!("Disconnected by portal: {reason}")));
+                }
+                Poll::Ready(phoenix_channel::Event::TokenExpired) => {
+                    return Poll::Ready(Err(anyhow!("Token expired")));
+                }
+                Poll::Ready(phoenix_channel::Event::Offline { req_id }) => {
+                    todo!()
+                }
+                Poll::Ready(phoenix_channel::Event::Disabled { req_id }) => {
+                    todo!()
+                }
+                Poll::Pending => {}
+            }
+
+            if self.print_stats_timer.poll_tick(cx).is_ready() {
+                tracing::debug!(target: "tunnel_state", stats = ?self.tunnel.stats());
+                continue;
+            }
+
+            return Poll::Pending;
+        }
+    }
+}
+
+// pub async fn request_log_upload_url(&mut self) {
+//     tracing::info!("Requesting log upload URL from portal");
+
+//     let _ = self
+//         .phoenix_channel
+//         .send(EgressMessages::CreateLogSink {})
+//         .await;
+// }
+
+// Errors here means we need to disconnect
+// #[tracing::instrument(level = "trace", skip(self))]
+// pub async fn handle_error(
+//     &mut self,
+//     reply_error: ChannelError,
+//     reference: Option<Reference>,
+//     topic: String,
+// ) -> Result<()> {
+//     match (reply_error, reference) {
+//         (ChannelError::ErrorReply(ErrorInfo::Offline), Some(reference)) => {
+//             let Ok(resource_id) = reference.parse::<ResourceId>() else {
+//                 tracing::warn!("The portal responded with an Offline error. Is the Resource associated with any online Gateways or Relays?");
+//                 return Ok(());
+//             };
+//             // TODO: Rate limit the number of attempts of getting the relays before just trying a local network connection
+//             self.tunnel.cleanup_connection(resource_id);
```

Comment:
Similarly, I don't quite understand what is happening here. We seem to shove different data types into the `reference` parameter. I would like to change that.

If we need to track additional data we should have a `HashMap` of `OutboundRequestId` to whatever data we need to have from the request. (We could also a feature to `phoenix-channel` to track the request and re-emit it upon response / error).

Reply:
This is actually a long time left over, we no longer keep track of the resource id on the reference.

The only time we really use it, that I remember, is when we send a resource request to make sure that we only use the response for the latest one we made since we could've made many of them and the state on the portal could change if there's a hold up somewhere.

Reply:
To implement protocol properly its better to always do something like promises - always track the response id, because protocol is asynchronous. But it up to you guys to decide how you need to use it